File: src/main/java/com/github/ptrteixeira/nusports/view/MainView.java
package com.github.ptrteixeira.nusports.view;

import com.github.ptrteixeira.nusports.model.Match;
import com.github.ptrteixeira.nusports.model.Standing;
import javafx.beans.value.ChangeListener;
import javafx.event.EventHandler;
import javafx.geometry.Pos;
import javafx.scene.Node;
import javafx.scene.control.Button;
import javafx.scene.control.Tab;
import javafx.scene.control.TabPane;
import javafx.scene.input.MouseEvent;
import javafx.scene.layout.BorderPane;
import javafx.scene.layout.HBox;
import javafx.scene.layout.Pane;
import org.apache.logging.log4j.LogManager;
import org.apache.logging.log4j.Logger;

import java.util.List;

public class MainView implements ViewPresenter {
  private static final Logger logger = LogManager.getLogger();

  private final ScheduleTab scheduleTab;
  private final StandingsTab standingsTab;
  private final TabPane tabPane;
  private final Button reloadButton;

  private DisplayType displayType;

  public MainView() {
    this(new ScheduleTab(), new StandingsTab());
  }

  MainView(ScheduleTab scheduleTab, StandingsTab standingsTab) {
    this.scheduleTab = scheduleTab;
    this.scheduleTab.setId("standingsTab");

    this.standingsTab = standingsTab;
    this.standingsTab.setId("scheduleTab");

    this.displayType = DisplayType.SCHEDULE;

    this.tabPane = new TabPane();
    this.tabPane.setId("tabPane");

    this.reloadButton = new Button("Reload");
    reloadButton.setId("reload");
  }

  @Override
  public void registerTabSwitchCallback(ChangeListener<Tab> tabChangeListener) {
    this.tabPane.getSelectionModel()
        .selectedItemProperty()
        .addListener((observable, oldValue, newValue) -> {
          // This part has to do with how the UI changes in response to a changed tab
          if (this.displayType == DisplayType.STANDINGS) {
            this.displayType = DisplayType.SCHEDULE;
          } else {
            this.displayType = DisplayType.STANDINGS;
          }

          // This part has to do with how the Model changes in response to a changed tab
          tabChangeListener.changed(observable, oldValue, newValue);
        });
  }

  @Override
  public void registerReloadCallback(EventHandler<MouseEvent> reloadCallback) {
    this.reloadButton.setOnMouseClicked(reloadCallback);
  }

  @Override
  public void registerSportSelectionCallback(ChangeListener<String> selectionChangeListener) {
    ChangeListener<String> callback = (observable, oldValue, newValue) -> {
      this.scheduleTab.setSelectedSport(newValue);
      this.standingsTab.setSelectedSport(newValue);

      selectionChangeListener.changed(observable, oldValue, newValue);
    };

    this.scheduleTab.registerSportSelections(callback);
    this.standingsTab.registerSportSelections(callback);
  }

  @Override
  public void setSelectableSports(List<String> sports) {
    logger.info("Set list of selectable sports to {}", sports);
    this.scheduleTab.setSportSelections(sports);
    this.standingsTab.setSportSelections(sports);

  }

  @Override
  public void setScheduleContents(List<Match> scheduleContents) {
    logger.trace("Set contents of schedule table to {}", scheduleContents);
    this.scheduleTab.populateTable(scheduleContents);
  }

  @Override
  public void setStandingsContents(List<Standing> standingsContents) {
    logger.trace("Reset contents of standings tab to {}", standingsContents);
    this.standingsTab.populateTable(standingsContents);
  }

  @Override
  public DisplayType currentDisplayType() {
    return this.displayType;
  }

  @Override
  public void setCurrentDisplayType(DisplayType displayType) {
    logger.debug("Set current DisplayType to {}", displayType.toString());
    this.displayType = displayType;
  }

  @Override
  public void clearErrorText() {
    logger.trace("Cleared the error message in the UI");
    this.standingsTab.clearErrorText();
    this.scheduleTab.clearErrorText();
  }

  @Override
  public void setErrorText(String text) {
    logger.trace("Set error message on UI to \"{}\"", text);
    this.scheduleTab.setErrorText(text);
    this.standingsTab.setErrorText(text);
  }

  @Override
  public Pane createView() {
    BorderPane pane = new BorderPane();
    pane.setCenter(this.createCenter());
    pane.setBottom(this.createBottomBar());

    pane.setMinWidth(600);
    pane.setMinHeight(600);

    logger.info("Creating main view");
    return pane;
  }

  private Node createBottomBar() {
    HBox bottomBar = new HBox();
    bottomBar.setAlignment(Pos.CENTER_RIGHT);
    Button reload = this.reloadButton;

    bottomBar.getChildren().add(reload);

    logger.trace("Creating bottom bar for main view");
    return bottomBar;
  }

  private Node createCenter() {
    TabPane tabPane = this.tabPane;
    tabPane.getTabs()
        .addAll(this.scheduleTab, this.standingsTab);

    logger.trace("Creating center for main view");
    return tabPane;
  }
}

File: src/main/kotlin/com/github/ptrteixeira/nusports/view/MainView.kt
/*
 * Copyright (c) 2017 Peter Teixeira
 *
 * Permission is hereby granted, free of charge, to any person obtaining a copy
 * of this software and associated documentation files (the "Software"), to deal
 * in the Software without restriction, including without limitation the rights
 * to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
 * copies of the Software, and to permit persons to whom the Software is
 * furnished to do so, subject to the following conditions:
 *
 * The above copyright notice and this permission notice shall be included in all
 * copies or substantial portions of the Software.
 *
 * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
 * IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
 * FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
 * AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
 * LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
 * OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
 * SOFTWARE.
 */
package com.github.ptrteixeira.nusports.view

import com.github.ptrteixeira.nusports.presenter.MainController
import com.github.ptrteixeira.nusports.view.DisplayType.SCHEDULE
import com.github.ptrteixeira.nusports.view.DisplayType.STANDINGS
import javafx.beans.value.ChangeListener
import javafx.geometry.Pos
import javafx.scene.control.Tab
import javafx.scene.control.TabPane
import javafx.scene.control.TabPane.TabClosingPolicy.UNAVAILABLE
import tornadofx.View
import tornadofx.add
import tornadofx.borderpane
import tornadofx.bottom
import tornadofx.button
import tornadofx.center
import tornadofx.hbox
import tornadofx.tab
import tornadofx.tabpane

class MainView(private val controller : MainController) : View() {
    private val scheduleTab = ScheduleTab(controller.displayedSchedule)
    private val standingsTab = StandingTab(controller.displayedStandings)

    private var displayType: DisplayType = SCHEDULE
    private var currentSelection = "Baseball"

    override val root = borderpane {
        minWidth = 600.toDouble()
        minHeight = 600.toDouble()

        center {
            tabpane {
                tabClosingPolicy = UNAVAILABLE

                setTabChangeListener(changeTabs())

                tab("Schedule") {
                    addRoot(scheduleTab) {
                        setSportSelections(controller.getSelectableSports())
                        setSportSelectionCallback(changeSport())
                    }
                }
                tab("Standings") {
                    addRoot(standingsTab) {
                        setSportSelections(controller.getSelectableSports())
                        setSportSelectionCallback(changeSport())
                    }
                }
            }
        }

        bottom {
            hbox {
                alignment = Pos.CENTER_RIGHT

                button("Reload") {
                    setOnMouseClicked {
                        controller
                            .lookup(displayType, currentSelection, true)
                    }
                }
            }
        }
    }

    private fun changeTabs() : ChangeListener<Tab> {
        return ChangeListener { _, _, newValue ->
            displayType = when (newValue.text) {
                "Schedule" -> SCHEDULE
                "Standings" -> STANDINGS
                else -> throw IllegalStateException("Invalid tab title")
            }

            controller.lookup(displayType, currentSelection)
        }
    }

    private fun changeSport(): ChangeListener<String> {
        return ChangeListener { _, _, newValue ->
            currentSelection = newValue

            controller.lookup(displayType, currentSelection, clearOnFail = true)
            scheduleTab.setSelectedSport(currentSelection)
            standingsTab.setSelectedSport(currentSelection)
        }
    }
}

private fun TabPane.setTabChangeListener(op: ChangeListener<in Tab>) {
    this.selectionModel
        .selectedItemProperty()
        .addListener(op)
}

private fun Tab.addRoot(abstractViewTab: AbstractTab, op: AbstractTab.() -> Unit) {
    abstractViewTab.apply(op)

    this.add(abstractViewTab.root)
}
